File: javascript_code/practical/underground/UndergroundSystem.js
// 抖码算法，让算法学习变的简单有趣
// 作者：老汤


var UndergroundSystem = function() {
    this.startInfo = new Map()
    this.table = new Map()
};

/**
 * @param {number} id
 * @param {string} stationName
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkIn = function(id, stationName, t) {
    const startObj = {
        "station": stationName,
        "time": t
    }
    this.startInfo.set(id, startObj)
};

/**
 * @param {number} id
 * @param {string} stationName
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkOut = function(id, stationName, t) {
    // 拿到起点的信息
    const startObj = this.startInfo.get(id)

    // 计算花费的时间
    const key = `${startObj.station}-${stationName}`
    // 这里 sumAmount 是一个包含两个元素的数组
    // 第一个元素表示的是总时间，第二个元素表示的是乘客总数
    const sumAmount = this.table.has(key) ? this.table.get(key) : [0, 0]

    sumAmount[0] += (t - startObj.time) // 总时间
    sumAmount[1] += 1   // 乘客总数

    this.table.set(key, sumAmount)
};

/**
 * @param {string} startStation
 * @param {string} endStation
 * @return {number}
 */
UndergroundSystem.prototype.getAverageTime = function(startStation, endStation) {
    const key = `${startStation}-${endStation}`
    const sumAmount = this.table.get(key)
    return sumAmount[0] / sumAmount[1]
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * var obj = new UndergroundSystem()
 * obj.checkIn(id,stationName,t)
 * obj.checkOut(id,stationName,t)
 * var param_3 = obj.getAverageTime(startStation,endStation)
 */

File: python_code/practical/underground/UndergroundSystem.py
# 抖码算法，让算法学习变得简单有趣
# 作者：老汤


class Start:
    def __init__(self, station, time):
        self.station = station
        self.time = time


class SumAmount:
    def __init__(self, sum, amount):
        self.sum = sum
        self.amount = amount


class UndergroundSystem:
    def __init__(self):
        # 维护每个乘客的起始站信息
        self.startInfo = dict()
        # 维护两个站的乘客信息：所有的乘客数，以及所有乘客花的时间
        self.table = dict()

    # 编号为 id 的乘客在 t 时刻进入地铁站 stationName
    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.startInfo[id] = Start(stationName, t)

    # 编号为 id 的乘客在 t 时刻离开地铁站 stationName
    def checkOut(self, id: int, stationName: str, t: int) -> None:
        start = self.startInfo[id]
        # 计算花费的时间
        key = "{}-{}".format(start.station, stationName)
        sumAmount = self.table[key] if key in self.table else SumAmount(0, 0)
        sumAmount.sum += (t - start.time)
        sumAmount.amount += 1
        self.table[key] = sumAmount

    # 返回地铁站 startStation 到地铁站 endStation 的平均花费时间
    def getAverageTime(self, startStation: str, endStation: str) -> float:
        key = "{}-{}".format(startStation, endStation)
        sumAmount = self.table[key]
        return sumAmount.sum / sumAmount.amount